lib/endpoints/class-wp-rest-posts-terms-controller.php

```diff
@@ -0,0 +1,225 @@
+<?php
+
+class WP_REST_Posts_Terms_Controller extends WP_REST_Controller {
+
+	protected $post_type;
+
+	public function __construct( $post_type, $taxonomy ) {
+		$this->post_type = $post_type;
+		$this->taxonomy = $taxonomy;
+		$this->posts_controller = new WP_REST_Posts_Controller( $post_type );
+		$this->terms_controller = new WP_REST_Terms_Controller( $taxonomy );
+	}
+
+	/**
+	 * Register the routes for the objects of the controller.
+	 */
+	public function register_routes() {
+
+		$base = $this->posts_controller->get_post_type_base( $this->post_type );
+
+		register_rest_route( 'wp/v2', sprintf( '/%s/(?P<id>[\d]+)/terms/%s', $base, $this->taxonomy ), array(
+			array(
+				'methods'             => WP_REST_Server::READABLE,
+				'callback'            => array( $this, 'get_items' ),
+				'permission_callback' => array( $this, 'get_items_permissions_check' ),
+			),
+		) );
+
+		register_rest_route( 'wp/v2', sprintf( '/%s/(?P<id>[\d]+)/terms/%s/(?P<term_id>[\d]+)', $base, $this->taxonomy ), array(
+			array(
+				'methods'             => WP_REST_Server::READABLE,
+				'callback'            => array( $this, 'get_item' ),
+				'permission_callback' => array( $this, 'get_items_permissions_check' ),
+			),
+			array(
+				'methods'             => WP_REST_Server::CREATABLE,
+				'callback'            => array( $this, 'create_item' ),
+				'permission_callback' => array( $this, 'create_item_permissions_check' ),
+			),
+			array(
+				'methods'         => WP_REST_Server::DELETABLE,
+				'callback'        => array( $this, 'delete_item' ),
+				'permission_callback' => array( $this, 'create_item_permissions_check' ),
+			),
+		) );
+	}
+
+	/**
+	 * Get all the terms that are attached to a post
+	 *
+	 * @param WP_REST_Request $request Full details about the request
+	 * @return WP_Error|WP_REST_Response
+	 */
+	public function get_items( $request ) {
+
+		$post = get_post( $request['id'] );
+
+		$post_check = $this->posts_controller->get_item( $request );
+		if ( is_wp_error( $post_check ) ) {
+			return $post_check;
+		}
+
+		$valid_taxonomies = get_object_taxonomies( $post->post_type );
```

Comment:
Actually, turns out we don't need this now that the controller is a post_type / taxonomy pair. This checking is done already on route registration in `plugin.php`.